File: rust/xdp-receiver/src/receiver.rs
use std::os::fd::AsRawFd;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

use anyhow::{Context, Result};

use crate::config::Config;
use crate::shred_parser;
use crate::stats::Stats;

const ETH_HDR_LEN: usize = 14;
const IPV4_HDR_MIN_LEN: usize = 20;
const GRE_HDR_MIN_LEN: usize = 4;
const UDP_HDR_LEN: usize = 8;

/// Represents the AF_XDP socket and UMEM resources.
pub struct AfXdpReceiver {
    fill_queue: xsk_rs::FillQueue,
    rx_queue: xsk_rs::RxQueue,
    umem: xsk_rs::Umem,
    frame_descs: Vec<xsk_rs::FrameDesc>,
    frame_count: usize,
}

impl AfXdpReceiver {
    /// Create and bind an AF_XDP socket to the specified interface and RX queue.
    /// Returns the receiver and the socket's raw fd (for registering in XSKMAP).
    pub fn new(config: &Config) -> Result<(Self, i32)> {
        let frame_count = config.frame_count();
        let frame_size = config.xdp.frame_size as u32;

        // Configure UMEM
        let umem_config = xsk_rs::UmemConfig::builder()
            .frame_count(frame_count as u32)
            .frame_size(frame_size)
            .fill_queue_size(frame_count as u32)
            .comp_queue_size(frame_count as u32)
            .build()
            .context("invalid UMEM config")?;

        let (umem, frame_descs) = xsk_rs::Umem::new(umem_config, frame_count as u32, false)
            .context("failed to create UMEM")?;

        // Configure socket
        let socket_config = xsk_rs::SocketConfig::builder()
            .rx_queue_size(frame_count as u32)
            .tx_queue_size(0) // RX only
            .build()
            .context("invalid socket config")?;

        let iface = xsk_rs::Interface::new(&config.network.physical_interface)
            .with_context(|| {
                format!(
                    "interface '{}' not found",
                    config.network.physical_interface
                )
            })?;

        let (_tx_queue, rx_queue, fq_cq) = xsk_rs::Socket::new(
            socket_config,
            &umem,
            &iface,
            config.xdp.rx_queue,
        )
        .context("failed to create AF_XDP socket")?;

        let (fill_queue, _comp_queue) = fq_cq
            .context("fill/completion queues not available (shared UMEM?)")?;

        let raw_fd = rx_queue.as_raw_fd();

        let receiver = Self {
            fill_queue,
            rx_queue,
            umem,
            frame_descs,
            frame_count,
        };

        Ok((receiver, raw_fd))
    }

    /// Populate the fill ring with all available frame addresses.
    pub fn fill_ring(&mut self) -> Result<()> {
        let submitted = unsafe { self.fill_queue.produce(&self.frame_descs) };
        eprintln!("Submitted {} frames to fill ring", submitted);
        Ok(())
    }

    /// Run the receive loop. Blocks until `shutdown` is set.
    pub fn run(
        &mut self,
        config: &Config,
        stats: Arc<RwLock<Stats>>,
        shutdown: Arc<AtomicBool>,
        ebpf: &aya::Ebpf,
    ) -> Result<()> {
        let poll_timeout_ms = 100;
        let mut xdp_stats_interval = Instant::now();
        let xdp_stats_period = Duration::from_secs(1);

        eprintln!(
            "AF_XDP receiver running on {} queue {}",
            config.network.physical_interface, config.xdp.rx_queue
        );

        while !shutdown.load(Ordering::Relaxed) {
            // Poll for received packets
            let frames_received = unsafe {
                self.rx_queue
                    .poll_and_consume(&mut self.frame_descs[..], poll_timeout_ms)
            };

            if frames_received == 0 {
                if xdp_stats_interval.elapsed() >= xdp_stats_period {
                    self.update_xdp_stats(&stats, ebpf);
                    xdp_stats_interval = Instant::now();
                }
                continue;
            }

            // Process received frames
            for i in 0..frames_received {
                let frame = &self.frame_descs[i];
                let pkt_data = unsafe { self.umem.data(frame) };
                self.process_packet(pkt_data, config, &stats);
            }

            // Return consumed frames to the fill ring
            let returned =
                unsafe { self.fill_queue.produce(&self.frame_descs[..frames_received]) };
            if returned < frames_received {
                stats.write().unwrap().afxdp_fill_starvation += 1;
            }
            stats.write().unwrap().afxdp_rx_fill_level =
                self.frame_count.saturating_sub(frames_received) + returned;

            // Periodically read BPF stats
            if xdp_stats_interval.elapsed() >= xdp_stats_period {
                self.update_xdp_stats(&stats, ebpf);
                xdp_stats_interval = Instant::now();
            }
        }

        eprintln!("AF_XDP receiver shutting down");
        Ok(())
    }

    /// Process a single received packet. Strips encapsulation headers and parses payload.
    fn process_packet(&self, pkt: &[u8], config: &Config, stats: &Arc<RwLock<Stats>>) {
        match Self::find_udp_payload(pkt) {
            Some((offset, port)) => {
                if port == config.network.heartbeat_port {
                    stats.write().unwrap().record_heartbeat();
                    return;
                }
                if port != config.network.shred_port {
                    return;
                }
                let payload = &pkt[offset..];
                match shred_parser::parse_shred(payload) {
                    Some(parsed) => {
                        let mut s = stats.write().unwrap();
                        s.record_shred(
                            parsed.slot,
                            parsed.is_data,
                            parsed.index,
                            parsed.fec_set_index,
                            parsed.signature,
                        );
                    }
                    None => {
                        stats.write().unwrap().record_parse_error();
                    }
                }
            }
            None => {
                stats.write().unwrap().record_parse_error();
            }
        }
    }

    /// Parse through Eth → outer IP → GRE → inner IP → UDP headers.
    /// Returns (payload_offset, udp_dst_port) if successful.
    fn find_udp_payload(pkt: &[u8]) -> Option<(usize, u16)> {
        if pkt.len() < ETH_HDR_LEN + IPV4_HDR_MIN_LEN {
            return None;
        }

        let mut offset = ETH_HDR_LEN;

        // Outer IPv4: read IHL
        let outer_ihl = ((pkt[offset] & 0x0F) as usize) * 4;
        if outer_ihl < IPV4_HDR_MIN_LEN || offset + outer_ihl > pkt.len() {
            return None;
        }
        // Check protocol == GRE (47)
        if pkt[offset + 9] != 47 {
            return None;
        }
        offset += outer_ihl;

        // GRE header
        if offset + GRE_HDR_MIN_LEN > pkt.len() {
            return None;
        }
        let gre_flags = u16::from_be_bytes([pkt[offset], pkt[offset + 1]]);
        let mut gre_len = GRE_HDR_MIN_LEN;
        if gre_flags & 0x8000 != 0 {
            gre_len += 4; // Checksum
        }
        if gre_flags & 0x2000 != 0 {
            gre_len += 4; // Key
        }
        if gre_flags & 0x1000 != 0 {
            gre_len += 4; // Sequence
        }
        offset += gre_len;

        // Inner IPv4
        if offset + IPV4_HDR_MIN_LEN > pkt.len() {
            return None;
        }
        let inner_ihl = ((pkt[offset] & 0x0F) as usize) * 4;
        if inner_ihl < IPV4_HDR_MIN_LEN || offset + inner_ihl > pkt.len() {
            return None;
        }
        if pkt[offset + 9] != 17 {
            return None; // Not UDP
        }
        offset += inner_ihl;

        // UDP header
        if offset + UDP_HDR_LEN > pkt.len() {
            return None;
        }
        let dst_port = u16::from_be_bytes([pkt[offset + 2], pkt[offset + 3]]);
        let payload_start = offset + UDP_HDR_LEN;

        Some((payload_start, dst_port))
    }

    fn update_xdp_stats(&self, stats: &Arc<RwLock<Stats>>, ebpf: &aya::Ebpf) {
        if let Ok((redirected, passed, errors)) = crate::xdp::read_xdp_stats(ebpf) {
            let mut s = stats.write().unwrap();
            s.update_xdp_counters(redirected, passed, errors);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Build a minimal GRE-encapsulated UDP packet for testing header parsing.
    fn build_gre_udp_packet(dst_port: u16) -> Vec<u8> {
        let mut pkt = Vec::new();

        // Ethernet header (14 bytes)
        pkt.extend_from_slice(&[0u8; 12]); // dst + src MAC
        pkt.extend_from_slice(&[0x08, 0x00]); // EtherType: IPv4

        // Outer IPv4 (20 bytes) - protocol 47 (GRE)
        pkt.push(0x45); // version=4, IHL=5
        pkt.extend_from_slice(&[0u8; 8]); // ToS, TotalLen, ID, Flags, TTL
        pkt.push(47); // Protocol: GRE
        pkt.extend_from_slice(&[0u8; 2]); // Header checksum
        pkt.extend_from_slice(&[10, 0, 0, 1]); // Src IP
        pkt.extend_from_slice(&[10, 0, 0, 2]); // Dst IP

        // GRE header (4 bytes, no optional fields)
        pkt.extend_from_slice(&[0x00, 0x00]); // Flags: none
        pkt.extend_from_slice(&[0x08, 0x00]); // Protocol: IPv4

        // Inner IPv4 (20 bytes) - protocol 17 (UDP)
        pkt.push(0x45); // version=4, IHL=5
        pkt.extend_from_slice(&[0u8; 8]); // ToS, TotalLen, etc
        pkt.push(17); // Protocol: UDP
        pkt.extend_from_slice(&[0u8; 2]); // Header checksum
        pkt.extend_from_slice(&[148, 51, 0, 1]); // Src IP
        pkt.extend_from_slice(&[233, 84, 178, 1]); // Dst IP (multicast)

        // UDP header (8 bytes)
        pkt.extend_from_slice(&[0x00, 0x00]); // Src port
        pkt.extend_from_slice(&dst_port.to_be_bytes()); // Dst port
        pkt.extend_from_slice(&[0x00, 0x00]); // Length
        pkt.extend_from_slice(&[0x00, 0x00]); // Checksum

        // Payload (dummy data)
        pkt.extend_from_slice(&[0xAA; 100]);

        pkt
    }

    #[test]
    fn test_find_udp_payload_shred_port() {
        let pkt = build_gre_udp_packet(7733);
        let result = AfXdpReceiver::find_udp_payload(&pkt);
        assert!(result.is_some());
        let (offset, port) = result.unwrap();
        assert_eq!(port, 7733);
        // Eth(14) + outerIP(20) + GRE(4) + innerIP(20) + UDP(8) = 66
        assert_eq!(offset, 66);
    }

    #[test]
    fn test_find_udp_payload_heartbeat_port() {
        let pkt = build_gre_udp_packet(5765);
        let result = AfXdpReceiver::find_udp_payload(&pkt);
        assert!(result.is_some());
        let (_, port) = result.unwrap();
        assert_eq!(port, 5765);
    }

    #[test]
    fn test_find_udp_payload_truncated() {
        let pkt = vec![0u8; 30]; // Too short
        assert!(AfXdpReceiver::find_udp_payload(&pkt).is_none());
    }

    #[test]
    fn test_find_udp_payload_gre_with_key() {
        let mut pkt = Vec::new();

        // Ethernet (14)
        pkt.extend_from_slice(&[0u8; 12]);
        pkt.extend_from_slice(&[0x08, 0x00]);

        // Outer IPv4 (20) - GRE
        pkt.push(0x45);
        pkt.extend_from_slice(&[0u8; 8]);
        pkt.push(47);
        pkt.extend_from_slice(&[0u8; 2]);
        pkt.extend_from_slice(&[10, 0, 0, 1]);
        pkt.extend_from_slice(&[10, 0, 0, 2]);

        // GRE with Key flag set (8 bytes total)
        pkt.extend_from_slice(&[0x20, 0x00]); // Flags: Key bit (0x2000)
        pkt.extend_from_slice(&[0x08, 0x00]); // Protocol: IPv4
        pkt.extend_from_slice(&[0x00, 0x00, 0x00, 0x01]); // Key value

        // Inner IPv4 (20) - UDP
        pkt.push(0x45);
        pkt.extend_from_slice(&[0u8; 8]);
        pkt.push(17);
        pkt.extend_from_slice(&[0u8; 2]);
        pkt.extend_from_slice(&[148, 51, 0, 1]);
        pkt.extend_from_slice(&[233, 84, 178, 1]);

        // UDP (8)
        pkt.extend_from_slice(&[0x00, 0x00]);
        pkt.extend_from_slice(&7733u16.to_be_bytes());
        pkt.extend_from_slice(&[0x00, 0x00, 0x00, 0x00]);

        pkt.extend_from_slice(&[0xBB; 50]);

        let result = AfXdpReceiver::find_udp_payload(&pkt);
        assert!(result.is_some());
        let (offset, port) = result.unwrap();
        assert_eq!(port, 7733);
        // Eth(14) + outerIP(20) + GRE(8) + innerIP(20) + UDP(8) = 70
        assert_eq!(offset, 70);
    }
}

File: go/xdp-receiver/receiver.go
package main

import (
	"context"
	"fmt"
	"net"
	"os"
	"sync"
	"time"
	"unsafe"

	"github.com/malbeclabs/edge-multicast-ref/go/internal/shred"
	"golang.org/x/sys/unix"
)

const (
	ethHdrLen      = 14
	ipv4HdrMinLen  = 20
	greHdrMinLen   = 4
	udpHdrLen      = 8
	pollTimeoutMs  = 100
	xdpStatsperiod = time.Second
)

// AfXdpSocket wraps a raw AF_XDP socket with its UMEM and ring buffers.
type AfXdpSocket struct {
	fd         int
	umem       []byte
	frameSize  int
	frameCount int

	// Ring buffer pointers (mmap'd)
	fillRing *xskRing
	rxRing   *xskRing
}

// xskRing represents an XSK ring buffer (fill or rx).
type xskRing struct {
	producer *uint32
	consumer *uint32
	descs    unsafe.Pointer // fill: *uint64 (addrs), rx: *unix.XDPDesc
	mask     uint32
	size     uint32
	cachedProd uint32
	cachedCons uint32
}

// xdpDesc matches the kernel's struct xdp_desc.
type xdpDesc struct {
	Addr    uint64
	Len     uint32
	Options uint32
}

// NewAfXdpSocket creates an AF_XDP socket, allocates UMEM, and sets up ring buffers.
func NewAfXdpSocket(cfg *Config) (*AfXdpSocket, error) {
	frameCount := cfg.FrameCount()
	frameSize := cfg.Xdp.FrameSize
	umemSize := frameCount * frameSize

	// Allocate UMEM region (page-aligned).
	umem, err := unix.Mmap(-1, 0, umemSize, unix.PROT_READ|unix.PROT_WRITE,
		unix.MAP_PRIVATE|unix.MAP_ANONYMOUS|unix.MAP_POPULATE)
	if err != nil {
		return nil, fmt.Errorf("mmap UMEM: %w", err)
	}

	// Create AF_XDP socket.
	fd, err := unix.Socket(unix.AF_XDP, unix.SOCK_RAW, 0)
	if err != nil {
		unix.Munmap(umem)
		return nil, fmt.Errorf("socket(AF_XDP): %w", err)
	}

	// Register UMEM.
	umemReg := unix.XDPUmemReg{
		Addr: uint64(uintptr(unsafe.Pointer(&umem[0]))),
		Len:  uint64(umemSize),
		Size: uint32(frameSize),
	}

	// Use power-of-2 ring sizes. Use frameCount if it's a power of 2,
	// otherwise find the largest power of 2 <= frameCount.
	ringSize := uint32(nearestPow2(frameCount))

	_, _, errno := unix.Syscall6(unix.SYS_SETSOCKOPT, uintptr(fd),
		unix.SOL_XDP, unix.XDP_UMEM_REG,
		uintptr(unsafe.Pointer(&umemReg)), unsafe.Sizeof(umemReg), 0)
	if errno != 0 {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("setsockopt XDP_UMEM_REG: %w", errno)
	}

	// Set ring sizes.
	if err := setsockoptInt(fd, unix.SOL_XDP, unix.XDP_UMEM_FILL_RING, int(ringSize)); err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("setsockopt FILL_RING: %w", err)
	}
	if err := setsockoptInt(fd, unix.SOL_XDP, unix.XDP_UMEM_COMPLETION_RING, int(ringSize)); err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("setsockopt COMPLETION_RING: %w", err)
	}
	if err := setsockoptInt(fd, unix.SOL_XDP, unix.XDP_RX_RING, int(ringSize)); err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("setsockopt RX_RING: %w", err)
	}

	// Get mmap offsets.
	offsets, err := getXdpMmapOffsets(fd)
	if err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("get mmap offsets: %w", err)
	}

	// mmap fill ring.
	fillMap, err := unix.Mmap(fd, unix.XDP_UMEM_PGOFF_FILL_RING,
		int(offsets.Fr.Desc)+int(ringSize)*8, // each fill desc is uint64
		unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED|unix.MAP_POPULATE)
	if err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		return nil, fmt.Errorf("mmap fill ring: %w", err)
	}

	fillRing := &xskRing{
		producer: (*uint32)(unsafe.Pointer(&fillMap[offsets.Fr.Producer])),
		consumer: (*uint32)(unsafe.Pointer(&fillMap[offsets.Fr.Consumer])),
		descs:    unsafe.Pointer(&fillMap[offsets.Fr.Desc]),
		mask:     ringSize - 1,
		size:     ringSize,
	}

	// mmap rx ring.
	rxMap, err := unix.Mmap(fd, unix.XDP_PGOFF_RX_RING,
		int(offsets.Rx.Desc)+int(ringSize)*int(unsafe.Sizeof(xdpDesc{})),
		unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED|unix.MAP_POPULATE)
	if err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		unix.Munmap(fillMap)
		return nil, fmt.Errorf("mmap rx ring: %w", err)
	}

	rxRing := &xskRing{
		producer: (*uint32)(unsafe.Pointer(&rxMap[offsets.Rx.Producer])),
		consumer: (*uint32)(unsafe.Pointer(&rxMap[offsets.Rx.Consumer])),
		descs:    unsafe.Pointer(&rxMap[offsets.Rx.Desc]),
		mask:     ringSize - 1,
		size:     ringSize,
	}

	// Bind socket.
	sa := unix.SockaddrXDP{
		Flags:   unix.XDP_USE_NEED_WAKEUP,
		Ifindex: uint32(mustIfaceIndex(cfg.Network.PhysicalInterface)),
		QueueID: cfg.Xdp.RxQueue,
	}
	if err := unix.Bind(fd, &sa); err != nil {
		unix.Close(fd)
		unix.Munmap(umem)
		unix.Munmap(fillMap)
		unix.Munmap(rxMap)
		return nil, fmt.Errorf("bind AF_XDP: %w", err)
	}

	sock := &AfXdpSocket{
		fd:         fd,
		umem:       umem,
		frameSize:  frameSize,
		frameCount: frameCount,
		fillRing:   fillRing,
		rxRing:     rxRing,
	}

	return sock, nil
}

// FD returns the socket file descriptor (for registering in XSKMAP).
func (s *AfXdpSocket) FD() int {
	return s.fd
}

// Close releases the AF_XDP socket and UMEM resources.
func (s *AfXdpSocket) Close() {
	unix.Close(s.fd)
	unix.Munmap(s.umem)
}

// FillRing submits all frames to the fill ring so the kernel can use them for RX.
func (s *AfXdpSocket) FillRing() {
	for i := 0; i < s.frameCount && i < int(s.fillRing.size); i++ {
		idx := *s.fillRing.producer & s.fillRing.mask
		fillDescs := (*[1 << 20]uint64)(s.fillRing.descs)
		fillDescs[idx] = uint64(i * s.frameSize)
		*s.fillRing.producer++
	}
	s.fillRing.cachedProd = *s.fillRing.producer
}

// ReceivePackets polls for received packets and returns the descriptors.
func (s *AfXdpSocket) ReceivePackets() []xdpDesc {
	cons := *s.rxRing.consumer
	prod := *s.rxRing.producer

	if cons == prod {
		return nil
	}

	descs := make([]xdpDesc, 0, prod-cons)
	rxDescs := (*[1 << 20]xdpDesc)(s.rxRing.descs)

	for cons != prod {
		idx := cons & s.rxRing.mask
		descs = append(descs, rxDescs[idx])
		cons++
	}

	*s.rxRing.consumer = cons
	return descs
}

// ReturnToFill returns consumed frame addresses to the fill ring.
func (s *AfXdpSocket) ReturnToFill(descs []xdpDesc) int {
	fillDescs := (*[1 << 20]uint64)(s.fillRing.descs)
	returned := 0
	for _, d := range descs {
		idx := *s.fillRing.producer & s.fillRing.mask
		fillDescs[idx] = d.Addr
		*s.fillRing.producer++
		returned++
	}
	return returned
}

// GetFrame returns the packet data from UMEM for a given descriptor.
func (s *AfXdpSocket) GetFrame(d xdpDesc) []byte {
	start := int(d.Addr)
	end := start + int(d.Len)
	if end > len(s.umem) {
		end = len(s.umem)
	}
	return s.umem[start:end]
}

// findUDPPayload parses through Eth -> outer IP -> GRE -> inner IP -> UDP headers.
// Returns (payload_offset, udp_dst_port, ok).
func findUDPPayload(pkt []byte) (int, uint16, bool) {
	if len(pkt) < ethHdrLen+ipv4HdrMinLen {
		return 0, 0, false
	}

	offset := ethHdrLen

	// Outer IPv4: read IHL
	outerIHL := int(pkt[offset]&0x0F) * 4
	if outerIHL < ipv4HdrMinLen || offset+outerIHL > len(pkt) {
		return 0, 0, false
	}
	// Check protocol == GRE (47)
	if pkt[offset+9] != 47 {
		return 0, 0, false
	}
	offset += outerIHL

	// GRE header
	if offset+greHdrMinLen > len(pkt) {
		return 0, 0, false
	}
	greFlags := uint16(pkt[offset])<<8 | uint16(pkt[offset+1])
	greLen := greHdrMinLen
	if greFlags&0x8000 != 0 {
		greLen += 4 // Checksum
	}
	if greFlags&0x2000 != 0 {
		greLen += 4 // Key
	}
	if greFlags&0x1000 != 0 {
		greLen += 4 // Sequence
	}
	offset += greLen

	// Inner IPv4
	if offset+ipv4HdrMinLen > len(pkt) {
		return 0, 0, false
	}
	innerIHL := int(pkt[offset]&0x0F) * 4
	if innerIHL < ipv4HdrMinLen || offset+innerIHL > len(pkt) {
		return 0, 0, false
	}
	if pkt[offset+9] != 17 { // Not UDP
		return 0, 0, false
	}
	offset += innerIHL

	// UDP header
	if offset+udpHdrLen > len(pkt) {
		return 0, 0, false
	}
	dstPort := uint16(pkt[offset+2])<<8 | uint16(pkt[offset+3])
	return offset + udpHdrLen, dstPort, true
}

// RunRecvLoop runs the AF_XDP receive loop, processing packets until ctx is done.
func RunRecvLoop(ctx context.Context, cfg *Config, xdpStats *XdpReceiverStats, mu *sync.RWMutex, handle *XdpHandle, sock *AfXdpSocket) error {
	// Fill the RX ring with all available frames.
	sock.FillRing()
	fmt.Fprintf(os.Stderr, "AF_XDP receiver running on %s queue %d\n",
		cfg.Network.PhysicalInterface, cfg.Xdp.RxQueue)

	pollFds := []unix.PollFd{
		{Fd: int32(sock.fd), Events: unix.POLLIN},
	}

	xdpStatsTimer := time.Now()

	for {
		select {
		case <-ctx.Done():
			fmt.Fprintln(os.Stderr, "AF_XDP receiver shutting down")
			return nil
		default:
		}

		// Poll for received packets.
		n, err := unix.Poll(pollFds, pollTimeoutMs)
		if err != nil {
			if err == unix.EINTR {
				continue
			}
			return fmt.Errorf("poll: %w", err)
		}
		if n == 0 {
			// Timeout: periodically read XDP stats.
			if time.Since(xdpStatsTimer) >= xdpStatsperiod {
				updateBpfStats(handle, xdpStats, mu)
				xdpStatsTimer = time.Now()
			}
			continue
		}

		// Receive packets from the RX ring.
		rxDescs := sock.ReceivePackets()
		if len(rxDescs) == 0 {
			continue
		}

		for _, desc := range rxDescs {
			pkt := sock.GetFrame(desc)
			processPacket(pkt, cfg, xdpStats, mu)
		}

		// Return consumed frames to fill ring.
		returned := sock.ReturnToFill(rxDescs)
		mu.Lock()
		if returned < len(rxDescs) {
			xdpStats.AfxdpFillStarvation++
		}
		xdpStats.AfxdpRxFillLevel = sock.frameCount - len(rxDescs) + returned
		mu.Unlock()

		// Periodically read BPF stats.
		if time.Since(xdpStatsTimer) >= xdpStatsperiod {
			updateBpfStats(handle, xdpStats, mu)
			xdpStatsTimer = time.Now()
		}
	}
}

// processPacket strips GRE encapsulation, parses the shred payload, and updates stats.
func processPacket(pkt []byte, cfg *Config, xdpStats *XdpReceiverStats, mu *sync.RWMutex) {
	offset, port, ok := findUDPPayload(pkt)
	if !ok {
		mu.Lock()
		xdpStats.RecordParseError()
		mu.Unlock()
		return
	}
	if port == cfg.Network.HeartbeatPort {
		mu.Lock()
		xdpStats.RecordHeartbeat()
		mu.Unlock()
		return
	}
	if port != cfg.Network.ShredPort {
		return
	}

	parsed, err := shred.ParseShred(pkt[offset:])
	if err != nil {
		mu.Lock()
		xdpStats.RecordParseError()
		mu.Unlock()
		return
	}
	mu.Lock()
	xdpStats.RecordShred(parsed.Slot, parsed.IsData, parsed.Index, parsed.FECSetIndex, parsed.Signature)
	mu.Unlock()
}

// updateBpfStats reads the XDP stats from the BPF map and updates the receiver stats.
func updateBpfStats(handle *XdpHandle, xdpStats *XdpReceiverStats, mu *sync.RWMutex) {
	redirected, passed, errors, err := ReadXdpStats(&handle.Objs)
	if err != nil {
		return
	}
	mu.Lock()
	xdpStats.UpdateXdpCounters(redirected, passed, errors)
	mu.Unlock()
}

// --- Helpers ---

func mustIfaceIndex(name string) int {
	iface, err := net.InterfaceByName(name)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Interface %q not found: %v\n", name, err)
		os.Exit(1)
	}
	return iface.Index
}

func nearestPow2(n int) int {
	if n <= 0 {
		return 1
	}
	v := 1
	for v < n {
		v <<= 1
	}
	// If v > n, go back one step (largest pow2 <= n).
	if v > n {
		v >>= 1
	}
	if v == 0 {
		v = 1
	}
	return v
}

func setsockoptInt(fd, level, opt, value int) error {
	v := uint32(value)
	_, _, errno := unix.Syscall6(unix.SYS_SETSOCKOPT, uintptr(fd),
		uintptr(level), uintptr(opt),
		uintptr(unsafe.Pointer(&v)), unsafe.Sizeof(v), 0)
	if errno != 0 {
		return errno
	}
	return nil
}

// xdpMmapOffsets and xdpRingOffset mirror the kernel structures.
type xdpRingOffset struct {
	Producer uint64
	Consumer uint64
	Desc     uint64
	Flags    uint64
}

type xdpMmapOffsets struct {
	Rx xdpRingOffset
	Tx xdpRingOffset
	Fr xdpRingOffset
	Cr xdpRingOffset
}

func getXdpMmapOffsets(fd int) (*xdpMmapOffsets, error) {
	var offsets xdpMmapOffsets
	size := uint32(unsafe.Sizeof(offsets))
	_, _, errno := unix.Syscall6(unix.SYS_GETSOCKOPT, uintptr(fd),
		unix.SOL_XDP, unix.XDP_MMAP_OFFSETS,
		uintptr(unsafe.Pointer(&offsets)), uintptr(unsafe.Pointer(&size)), 0)
	if errno != 0 {
		return nil, errno
	}
	return &offsets, nil
}
